aws/resource_aws_globalaccelerator_listener.go

```diff
@@ -0,0 +1,270 @@
+	conn := meta.(*AWSClient).globalacceleratorconn
+
+	listener, err := resourceAwsGlobalAcceleratorListenerRetrieve(conn, d.Id())
+
+	if err != nil {
+		return fmt.Errorf("Error reading Global Accelerator listener: %s", err)
+	}
+
+	if listener == nil {
+		log.Printf("[WARN] Global Accelerator listener (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	acceleratorArn, err := resourceAwsGlobalAcceleratorListenerParseAcceleratorArn(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	d.Set("accelerator_arn", acceleratorArn)
+	d.Set("client_affinity", listener.ClientAffinity)
+	d.Set("protocol", listener.Protocol)
+	d.Set("port_range", resourceAwsGlobalAcceleratorListenerFlattenPortRanges(listener.PortRanges))
+
+	return nil
+}
+
+func resourceAwsGlobalAcceleratorListenerParseAcceleratorArn(listenerArn string) (string, error) {
+	parts := strings.Split(listenerArn, "/")
+	if len(parts) < 4 {
+		return "", fmt.Errorf("Unable to parse accelerator ARN from %s", listenerArn)
+	}
+	return strings.Join(parts[0:2], "/"), nil
+}
+
+func resourceAwsGlobalAcceleratorListenerExpandPortRanges(portRanges []interface{}) []*globalaccelerator.PortRange {
+	out := make([]*globalaccelerator.PortRange, len(portRanges))
+
+	for i, raw := range portRanges {
+		portRange := raw.(map[string]interface{})
+		m := globalaccelerator.PortRange{}
+
+		m.FromPort = aws.Int64(int64(portRange["from_port"].(int)))
+		m.ToPort = aws.Int64(int64(portRange["to_port"].(int)))
+
+		out[i] = &m
+	}
+
+	return out
+}
+
+func resourceAwsGlobalAcceleratorListenerFlattenPortRanges(portRanges []*globalaccelerator.PortRange) []interface{} {
+	out := make([]interface{}, len(portRanges))
+
+	for i, portRange := range portRanges {
+		m := make(map[string]interface{})
+
+		m["from_port"] = *portRange.FromPort
+		m["to_port"] = *portRange.ToPort
+
+		out[i] = m
+	}
+
+	log.Printf("[DEBUG] Flatten port_range: %s", out)
+	return out
+}
+
+func resourceAwsGlobalAcceleratorListenerRetrieve(conn *globalaccelerator.GlobalAccelerator, listenerArn string) (*globalaccelerator.Listener, error) {
+	resp, err := conn.DescribeListener(&globalaccelerator.DescribeListenerInput{
+		ListenerArn: aws.String(listenerArn),
+	})
+
+	if err != nil {
+		if isAWSErr(err, globalaccelerator.ErrCodeListenerNotFoundException, "") {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return resp.Listener, nil
+}
+
+func resourceAwsGlobalAcceleratorListenerUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).globalacceleratorconn
+
+	opts := &globalaccelerator.UpdateListenerInput{
+		ListenerArn: aws.String(d.Id()),
+		Protocol:    aws.String(d.Get("protocol").(string)),
+		PortRanges:  resourceAwsGlobalAcceleratorListenerExpandPortRanges(d.Get("port_range").([]interface{})),
+	}
+
+	if v, ok := d.GetOk("client_affinity"); ok {
+		opts.ClientAffinity = aws.String(v.(string))
```

Comment:
Since we are switching `client_affinity` to have a default, this can be moved up into the above struct. 👍